api/internal/localizer/localizer_test.go

```diff
@@ -113,5 +160,161 @@ resources:
+components:
+- epsilon
+kind: Component
+`,
+			"delta/epsilon/kustomization.yaml": `apiVersion: kustomize.config.k8s.io/v1alpha1
+kind: Component
+namespace: kustomize-namespace
+`,
+		},
+		"multiple": {
+			"kustomization.yaml": `apiVersion: kustomize.config.k8s.io/v1beta1
+components:
+- delta
+- delta/epsilon
+kind: Kustomization
+`,
+			"delta/kustomization.yaml": `apiVersion: kustomize.config.k8s.io/v1alpha1
+kind: Component
+namespace: kustomize-namespace
+`,
+			"delta/epsilon/kustomization.yaml": `apiVersion: kustomize.config.k8s.io/v1alpha1
+buildMetadata:
+- managedByLabel
+kind: Component
+`,
+		},
+		"localize_directory_name_when_absent": {
+			"kustomization.yaml": fmt.Sprintf(`apiVersion: kustomize.config.k8s.io/v1beta1
+components:
+- %s
+kind: Kustomization
+`, LocalizeDir),
+			fmt.Sprintf("%s/kustomization.yaml", LocalizeDir): `apiVersion: kustomize.config.k8s.io/v1alpha1
+kind: Component
+namespace: kustomize-namespace
+`,
+		},
+	} {
+		t.Run(name, func(t *testing.T) {
+			fSys := makeMemoryFs(t)
+			addFiles(t, fSys, "/alpha/beta/gamma", files)
+
+			lclzr := createLocalizer(t, fSys, "/alpha/beta/gamma", "/alpha/beta", "/dst")
+			err := lclzr.Localize()
+			require.NoError(t, err)
+
+			fSysExpected := makeMemoryFs(t)
+			addFiles(t, fSysExpected, "/alpha/beta/gamma", files)
+			addFiles(t, fSysExpected, "/dst/gamma", files)
+			checkFSys(t, fSysExpected, fSys)
+		})
+	}
+}
+
+func TestLocalizeDirCleanedSibling(t *testing.T) {
+	fSys := makeMemoryFs(t)
+	kustAndComponents := map[string]string{
+		"beta/gamma/kustomization.yaml": `apiVersion: kustomize.config.k8s.io/v1beta1
+components:
+- delta/../../../../a/b/../../alpha/beta/sibling
+kind: Kustomization`,
+		"beta/sibling/kustomization.yaml": `apiVersion: kustomize.config.k8s.io/v1alpha1
+kind: Component
+namespace: kustomize-namespace
+`,
+	}
+	addFiles(t, fSys, "/alpha", kustAndComponents)
+
+	lclzr := createLocalizer(t, fSys, "/alpha/beta/gamma", "/alpha", "/alpha/beta/dst")
+	err := lclzr.Localize()
+	require.NoError(t, err)
+
+	fSysExpected := makeMemoryFs(t)
+	addFiles(t, fSysExpected, "/alpha", kustAndComponents)
+	cleanedFiles := map[string]string{
+		"beta/gamma/kustomization.yaml": `apiVersion: kustomize.config.k8s.io/v1beta1
+components:
+- ../sibling
+kind: Kustomization
+`,
+		"beta/sibling/kustomization.yaml": kustAndComponents["beta/sibling/kustomization.yaml"],
+	}
+	addFiles(t, fSysExpected, "/alpha/beta/dst", cleanedFiles)
+	checkFSys(t, fSysExpected, fSys)
+}
+
+func TestLocalizeComponents(t *testing.T) {
+	fSys := makeMemoryFs(t)
+	kustAndComponents := map[string]string{
+		"kustomization.yaml": `apiVersion: kustomize.config.k8s.io/v1beta1
+components:
+- a
+- alpha
+kind: Kustomization
+`,
+		"a/kustomization.yaml": `apiVersion: kustomize.config.k8s.io/v1alpha1
+kind: Component
+patches:
+- patch: |-
+    - op: replace
+      path: /spec/template/spec/containers/0/name
+      value: my-nginx
+replacements:
```

Comment:
Is there any particular reason to add patches and replacements and replicas to these component files? I had a similar question for some of the tests in https://github.com/kubernetes-sigs/kustomize/pull/4865 as well.

I think unless we are specifically testing a particular field, we should try to keep the test kustomization files minimal, maybe just a namePrefix or something. Otherwise it adds noise to the test.

Reply:
Fixed.

I was trying to test that different fields in a `Component` are all copied as an extra precaution. I see your point about the clutter. The extra fields aren't strictly necessary because the logic used to process `Component`s is the same as that used for `Kustomization`s, and we've tested that already.